src/optimisation/Optimisation.jl

```diff
@@ -139,186 +154,364 @@ function (f::OptimLogDensity)(F, G, z)
+    return ModeResult(log_density, solution)
 end
 
-function _optim_objective(model::DynamicPPL.Model, ::MLE,  ::constrained_space{false})
-    ctx = OptimizationContext(DynamicPPL.LikelihoodContext())
-    obj = OptimLogDensity(model, ctx)
-    
-    obj = transform!!(obj)
-    init = Init(obj.varinfo, model, constrained_space{false}())
-    t = ParameterTransform(obj.varinfo, model, constrained_space{true}())
-    
-    return (obj=obj, init = init, transform=t)
+# If no solver is given.
+function estimate_mode(
+    model::DynamicPPL.Model,
+    estimator::ModeEstimator,
+    initial_params::Union{AbstractVector,Nothing},
+    args...;
+    kwargs...
+)
+    return estimate_mode(model, estimator, initial_params, nothing, args...; kwargs...)
 end
 
-function _optim_objective(model::DynamicPPL.Model, ::MLE, ::constrained_space{true})
-    ctx = OptimizationContext(DynamicPPL.LikelihoodContext())
-    obj = OptimLogDensity(model, ctx)
-  
-    init = Init(obj.varinfo, model, constrained_space{true}())
-    t = ParameterTransform(obj.varinfo, model, constrained_space{true}())
-    
-    return (obj=obj, init = init, transform=t)
+# If no initial value is given.
+function estimate_mode(
+    model::DynamicPPL.Model,
+    estimator::ModeEstimator,
+    solver,
+    args...;
+    kwargs...
+)
+    return estimate_mode(model, estimator, nothing, solver, args...; kwargs...)
 end
 
-function optim_objective(model::DynamicPPL.Model, estimator::Union{MLE, MAP}; constrained::Bool=true)
-    return _optim_objective(model, estimator, constrained_space{constrained}())
+# If no solver or initial value is given.
+function estimate_mode(
+    model::DynamicPPL.Model,
+    estimator::ModeEstimator,
+    args...;
+    kwargs...
+)
+    return estimate_mode(model, estimator, nothing, nothing, args...; kwargs...)
 end
 
+"""
+    maximum_a_posteriori(
+        model::DynamicPPL.Model,
+        [initial_params::Union{AbstractVector,Nothing}],
+        [solver];
+        <keyword arguments>
+    )
 
-function optim_function(
-    model::Model,
-    estimator::Union{MLE, MAP};
-    constrained::Bool=true,
-    adtype::Union{Nothing, AbstractADType}=NoAD(),
-)
-    if adtype === nothing
-        Base.depwarn("the use of `adtype=nothing` is deprecated, please use `adtype=SciMLBase.NoAD()`", :optim_function)
-    end
+Find the maximum a posteriori estimate of a model.
 
-    obj, init, t = optim_objective(model, estimator; constrained=constrained)
-    
-    l(x, _) = obj(x)
-    f = if adtype isa AbstractADType && adtype !== NoAD()
-        OptimizationFunction(l, adtype)
-    else
-        OptimizationFunction(
-            l;
-            grad = (G,x,p) -> obj(nothing, G, x),
-        )
-    end
-    
-    return (func=f, init=init, transform = t)
+This is a convenience function that calls `estimate_mode` with `MAP()` as the estimator.
+Please see the documentation of [`estimate_mode`](@ref) for more details.
```

Comment:
`estimate_mode` is no longer exported right?
```suggestion
Please see the documentation of [`Turing.Optimisation.estimate_mode`](@ref) for more details.
```

Reply:
Good catch, fixed.